pkg/dns/dns.go

```diff
@@ -0,0 +1,417 @@
+	return json
+}
+
+func (kd *KubeDNS) setServicesStore() {
+	// Returns a cache.ListWatch that gets all changes to services.
+	serviceWatch := kcache.NewListWatchFromClient(kd.kubeClient, "services", kapi.NamespaceAll, kselector.Everything())
+	kd.servicesStore, kd.serviceController = kframework.NewInformer(
+		serviceWatch,
+		&kapi.Service{},
+		resyncPeriod,
+		kframework.ResourceEventHandlerFuncs{
+			AddFunc:    kd.newService,
+			DeleteFunc: kd.removeService,
+			UpdateFunc: kd.updateService,
+		},
+	)
+}
+
+func (kd *KubeDNS) setEndpointsStore() {
+	// Returns a cache.ListWatch that gets all changes to endpoints.
+	endpointsWatch := kcache.NewListWatchFromClient(kd.kubeClient, "endpoints", kapi.NamespaceAll, kselector.Everything())
+	kd.endpointsStore, kd.eController = kframework.NewInformer(
+		endpointsWatch,
+		&kapi.Endpoints{},
+		resyncPeriod,
+		kframework.ResourceEventHandlerFuncs{
+			AddFunc: kd.handleEndpointAdd,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				// TODO: Avoid unwanted updates.
+				kd.handleEndpointAdd(newObj)
+			},
+		},
+	)
+}
+
+func (kd *KubeDNS) newService(obj interface{}) {
+	if service, ok := obj.(*kapi.Service); ok {
+		// if ClusterIP is not set, a DNS entry should not be created
+		if !kapi.IsServiceIPSet(service) {
+			kd.newHeadlessService(service)
+			return
+		}
+		if len(service.Spec.Ports) == 0 {
+			glog.Info("Unexpected service with no ports, this should not have happend: %v", service)
+		}
+		kd.newPortalService(service)
+	}
+}
+
+func (kd *KubeDNS) removeService(obj interface{}) {
+	if s, ok := obj.(*kapi.Service); ok {
+		subCachePath := append(kd.domainPath, serviceSubdomain, s.Namespace, s.Name)
+		kd.cache.DeletePath(subCachePath...)
+	}
+}
+
+func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
+	kd.newService(newObj)
+}
+
+func (kd *KubeDNS) handleEndpointAdd(obj interface{}) {
+	if e, ok := obj.(*kapi.Endpoints); ok {
+		kd.addDNSUsingEndpoints(e)
+	}
+}
+
+func (kd *KubeDNS) addDNSUsingEndpoints(e *kapi.Endpoints) error {
+	svc, err := kd.getServiceFromEndpoints(e)
+	if err != nil {
+		return err
+	}
+	if svc == nil || kapi.IsServiceIPSet(svc) {
+		// No headless service found corresponding to endpoints object.
+		return nil
+	}
+	return kd.generateRecordsForHeadlessService(e, svc)
+}
+
+func (kd *KubeDNS) getServiceFromEndpoints(e *kapi.Endpoints) (*kapi.Service, error) {
+	key, err := kcache.MetaNamespaceKeyFunc(e)
+	if err != nil {
+		return nil, err
+	}
+	obj, exists, err := kd.servicesStore.GetByKey(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get service object from services store - %v", err)
+	}
+	if !exists {
+		glog.V(1).Infof("could not find service for endpoint %q in namespace %q", e.Name, e.Namespace)
+		return nil, nil
+	}
+	if svc, ok := obj.(*kapi.Service); ok {
+		return svc, nil
+	}
+	return nil, fmt.Errorf("got a non service object in services store %v", obj)
+}
+
+func (kd *KubeDNS) newPortalService(service *kapi.Service) {
```

Comment:
For context: https://github.com/openshift/origin/blob/master/pkg/dns/serviceresolver.go#L75